tests/integrated/test_suite

```diff
@@ -100,42 +206,80 @@ test_type = "Making" if args.make else "Running"
 
 print("======= {} {} {} tests ========".format(test_type, len(tests), savepath.split("/")[-1]))
 
+avail = num_jobs
+tester = sorted([Test(t, command) for t in tests],
+                key=lambda x: x.cost, reverse=True)
+cost_remain = sum(x.cost for x in tester)
+if js_read:
+    js_tokens = os.read(js_read,cost_remain-1)
+    num_jobs = len(js_tokens)+1
+
+torun=[i for i in range(len(tests))]
+running=[]
+
 longest = max([len(s) for s in tests])
 output = '%-'+str(longest+1)+'s'
-for t in tests:
-    os.chdir(t)
-
-    print(output % t, end='')
-    sys.stdout.flush()  # Make sure name is flushed
-
-    # Check requirements
-    req_met, req_expr = requirements.check("./runtest")
-    if not req_met:
-        print("S - {0} => False".format(req_expr))
-        sys.stdout.flush()
-        os.chdir(savepath)
-        continue
-
-    start_time_ = time.time()
-    # Run test, piping stdout so it is not sent to console
-    status, out = shell(command, pipe=True)
-
-    if status != 0:
-        # ❌ Failed
-        print(u"\u274C", end='')  # No newline
-        failed.append((t, out))
+
+while len(torun)+len(running):
+    if avail:
+        # See whether we can start one
+        for i in torun:
+            if tester[i].cost <= avail:
+                torun.remove(i)
```

Comment:
Modifying lists inside loops usually best avoided as it can be the cause of subtle bugs.

Maybe loop over `testers` directly, and build up a list of what to remove, and actually remove them after the loop?

```python
if avail:
  tests_to_start = [test for test in tester if test.cost <= avail]

for job in tests_to_start:
  tester.remove(job)
  job.start()
  running.append(job)
  avail -= job.cost
  cost_remain -= job.cost
```

Reply:
I think taking a copy with `[:]` should prevent the issue as well?

As is, it wouldn't work because we need to modify `avail`(`avail_threads`) to make sure the remaining threads are also sufficient for the next test.